middleware/proxy/grpc.go

```diff
@@ -0,0 +1,70 @@
+package proxy
+
+import (
+	"crypto/tls"
+	"log"
+
+	"github.com/miekg/coredns/middleware/proxy/pb"
+	"github.com/miekg/coredns/request"
+
+	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"golang.org/x/net/context"
+
+)
+
+type grpcClient struct {
+	dialOpt grpc.DialOption
+	clients	map[string]pb.DnsServiceClient
+	upstream *staticUpstream
+}
+
+func newGrpcClient(tls *tls.Config, u *staticUpstream) *grpcClient {
+	g := &grpcClient{upstream: u}
+
+	if tls == nil {
+		g.dialOpt = grpc.WithInsecure()
+	} else {
+		g.dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tls))
+	}
+	g.clients = map[string]pb.DnsServiceClient{}
+
+	return g
+}
+
+func (g *grpcClient) Exchange(addr string, state request.Request) (*dns.Msg, error) {
+	msg, err := state.Req.Pack()
+	if err != nil {
+		return nil, err
+	}
+
+	reply, err := g.clients[addr].Query(context.TODO(), &pb.DnsPacket{Msg: msg})
+	if err != nil {
+		return nil, err
+	}
+	d := new(dns.Msg)
+	err = d.Unpack(reply.Msg)
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (g *grpcClient) Protocol() string { return "grpc" }
+
+func (g *grpcClient) OnShutdown(p *Proxy) error {
+	return nil
+}
+
+func (g *grpcClient) OnStartup(p *Proxy) error {
+	for _, host := range g.upstream.Hosts {
+		conn, err := grpc.Dial(host.Name, g.dialOpt)
+		if err != nil {
+			log.Printf("[WARNING] Skipping gRPC host '%s' due to Dial error: '%s'", host, err)
+		} else {
+			g.clients[host.Name] = pb.NewDnsServiceClient(conn)
```

Comment:
Yeah, I was thinking since it's in the startup phase it will only be called once, but I'll double-check.

Haven't figured out the redial and health check yet. I know the gRPC code will retry during dial but not sure if it will redial after a connection is lost.